lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -60,11 +66,51 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
 
   using OpConversionPattern::OpConversionPattern;
 
+  // Convert add lowers a poly.add operation to arith operations. A poly.add
+  // operation is defined within the polynomial ring. Coefficients are added
+  // element-wise as elements of the ring, so they are performed modulo the
+  // coefficient modulus.
+  //
+  // Modular addition is performed by sign extending both arguments and
+  // performing (N+1)-bit addition. Then, if z = (x + y) is the (N+1)-bit sum,
+  // we return
+  //   truncate(z >= mod ? z - mod : z)
   LogicalResult matchAndRewrite(
       AddOp op, OpAdaptor adaptor,
       ConversionPatternRewriter &rewriter) const override {
-    // TODO(https://github.com/google/heir/issues/104): implement
-    return failure();
+    auto type = adaptor.getLhs().getType();
+    auto elementTy =
+        type.cast<RankedTensorType>().getElementType().cast<IntegerType>();
+    auto extType = IntegerType::get(getContext(), elementTy.getWidth() + 1,
+                                    elementTy.getSignedness());
+    auto extShapeType = type.cast<ShapedType>().clone(extType);
+
+    mlir::Value extLhsOp = rewriter.create<arith::ExtUIOp>(
```

Comment:
Assuming optimizations will stay here: since you know the modulus and the element type bit width statically, you could check in advance whether there is a risk of overflow, and if not skip the sign extensions.